runtime/parachains/src/inclusion_inherent.rs

```diff
@@ -127,6 +129,47 @@ decl_module! {
 	}
 }
 
+/// Limit the number of backed candidates processed in order to stay within block weight limits.
+///
+/// Use a configured assumption about the weight required to process a backed candidate and the
+/// current block weight as of the execution of this function to ensure that we don't overload
+/// the block with candidate processing.
+///
+/// If the backed candidates exceed the available block weight remaining, then skips all of them.
+/// This is somewhat less desirable than attempting to fit some of them, but is more fair in the
+/// even that we can't trust the provisioner to provide a fair / random ordering of candidates.
+fn limit_backed_candidates<T: Config>(
+	backed_candidates: Vec<BackedCandidate<T::Hash>>,
+) -> Vec<BackedCandidate<T::Hash>> {
+	const BACKED_CANDIDATE_WEIGHT_ASSUMPTION: usize = 10_000;
+
+	let block_weight_remaining = <T as frame_system::Config>::MaximumBlockWeight::get()
+		.saturating_sub(frame_system::Module::<T>::block_weight().total());
+
+	if backed_candidates.len() * BACKED_CANDIDATE_WEIGHT_ASSUMPTION > block_weight_remaining as usize {
```

Comment:
With my limited knowledge, what happens to the inherents if they don't make it here? That would imply the candidates never make?
The threshold also seems to be very coarse. I think most of these points were already discussed in the comments.

Reply:
If they don't get included, then they'll be retried later. This is analogous to discarding the candidate as we do [here](https://github.com/paritytech/polkadot/blob/7fb1e57d7a90badbbe0b8098443b9da85d3f4da0/runtime/parachains/src/inclusion_inherent.rs#L141-L145) in `master` already.